File: classes/parte-4-datos/064-repositorio-y-dominio/implementaciones/sqlalchemy/dominio.py
"""EL DOMINIO.

Este archivo no importa SQLAlchemy, ni FastAPI, ni nada que sepa de bases de
datos. Es Python corriente, y por eso las reglas de mas abajo se pueden ejecutar
en una prueba en milisegundos — sin servidor, sin esquema y sin limpiar tablas
entre casos.

La ruta `/dominio` lo lee y comprueba, SOBRE EL TEXTO, que esa afirmacion es
cierta. Prometerlo en un README no cuesta nada; comprobarlo, si.
"""
from typing import Any


class ReglaRota(Exception):
    def __init__(self, codigo: str) -> None:
        super().__init__(codigo)
        self.codigo = codigo


class Tarea:
    def __init__(self, id_tarea: int, titulo: str, hecha: bool = False) -> None:
        self.id = id_tarea
        self.titulo = titulo
        self.hecha = hecha

    def terminar(self) -> None:
        self.hecha = True


class Proyecto:
    """El proyecto es la RAIZ: nadie toca una tarea sin pasar por el.

    Esa es la razon de que las tres reglas puedan vivir aqui. Si el resto del
    codigo pudiera anadir tareas por su cuenta, «no se anaden tareas a un
    proyecto cerrado» seria una recomendacion en lugar de una regla.
    """

    def __init__(
        self,
        id_proyecto: int,
        nombre: str,
        cerrado: bool = False,
        tareas: list[Tarea] | None = None,
    ) -> None:
        self.id = id_proyecto
        self.nombre = nombre
        self.cerrado = cerrado
        self.tareas: list[Tarea] = tareas or []

    def anadir_tarea(self, id_tarea: int, titulo: str) -> Tarea:
        """REGLA 2 y REGLA 3."""
        if self.cerrado:
            raise ReglaRota("PROYECTO_CERRADO")
        if any(t.titulo == titulo for t in self.tareas):
            raise ReglaRota("TITULO_REPETIDO")
        tarea = Tarea(id_tarea, titulo)
        self.tareas.append(tarea)
        return tarea

    def cerrar(self) -> None:
        """REGLA 1."""
        if self.pendientes() > 0:
            raise ReglaRota("QUEDAN_PENDIENTES")
        self.cerrado = True

    def terminar_tarea(self, id_tarea: int) -> Tarea:
        for tarea in self.tareas:
            if tarea.id == id_tarea:
                tarea.terminar()
                return tarea
        raise ReglaRota("NO_EXISTE")

    def pendientes(self) -> int:
        return sum(1 for t in self.tareas if not t.hecha)

    def salida(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "nombre": self.nombre,
            "cerrado": self.cerrado,
            "tareas": len(self.tareas),
            "pendientes": self.pendientes(),
        }

File: classes/parte-4-datos/064-repositorio-y-dominio/implementaciones/hibernate/src/main/java/labs/Dominio.java
package labs;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * EL DOMINIO.
 *
 * Este archivo no importa Hibernate, ni JPA, ni Spring. Es Java corriente, y por
 * eso las reglas de mas abajo se pueden ejecutar en una prueba en milisegundos —
 * sin servidor, sin esquema y sin limpiar tablas entre casos.
 *
 * La ruta `/dominio` lo lee y comprueba, SOBRE SUS IMPORTS, que esa afirmacion
 * es cierta. Prometerlo en un README no cuesta nada; comprobarlo, si.
 */
public final class Dominio {

    private Dominio() {
    }

    public static class ReglaRota extends RuntimeException {
        public final String codigo;

        public ReglaRota(String codigo) {
            super(codigo);
            this.codigo = codigo;
        }
    }

    public static final class Tarea {
        public final long id;
        public String titulo;
        public boolean hecha;

        public Tarea(long id, String titulo, boolean hecha) {
            this.id = id;
            this.titulo = titulo;
            this.hecha = hecha;
        }

        public void terminar() {
            this.hecha = true;
        }
    }

    /**
     * El proyecto es la RAIZ: nadie toca una tarea sin pasar por el.
     *
     * Esa es la razon de que las tres reglas puedan vivir aqui. Si el resto del
     * codigo pudiera anadir tareas por su cuenta, «no se anaden tareas a un
     * proyecto cerrado» seria una recomendacion en lugar de una regla.
     */
    public static final class Proyecto {
        public final long id;
        public String nombre;
        public boolean cerrado;
        public final List<Tarea> tareas;

        public Proyecto(long id, String nombre, boolean cerrado, List<Tarea> tareas) {
            this.id = id;
            this.nombre = nombre;
            this.cerrado = cerrado;
            this.tareas = new ArrayList<>(tareas);
        }

        public Proyecto(long id, String nombre) {
            this(id, nombre, false, List.of());
        }

        /** REGLA 2 y REGLA 3. */
        public Tarea anadirTarea(long idTarea, String titulo) {
            if (cerrado) {
                throw new ReglaRota("PROYECTO_CERRADO");
            }
            if (tareas.stream().anyMatch(t -> t.titulo.equals(titulo))) {
                throw new ReglaRota("TITULO_REPETIDO");
            }
            Tarea tarea = new Tarea(idTarea, titulo, false);
            tareas.add(tarea);
            return tarea;
        }

        /** REGLA 1. */
        public void cerrar() {
            if (pendientes() > 0) {
                throw new ReglaRota("QUEDAN_PENDIENTES");
            }
            cerrado = true;
        }

        public Tarea terminarTarea(long idTarea) {
            for (Tarea tarea : tareas) {
                if (tarea.id == idTarea) {
                    tarea.terminar();
                    return tarea;
                }
            }
            throw new ReglaRota("NO_EXISTE");
        }

        public int pendientes() {
            return (int) tareas.stream().filter(t -> !t.hecha).count();
        }

        public Map<String, Object> salida() {
            return Map.of(
                    "id", (int) id,
                    "nombre", nombre,
                    "cerrado", cerrado,
                    "tareas", tareas.size(),
                    "pendientes", pendientes());
        }
    }

    /** Lo unico que el dominio necesita. Cuatro metodos, y ninguno menciona SQL. */
    public interface Repositorio {
        Proyecto porId(long id);

        Proyecto guardar(Proyecto proyecto);

        long siguienteIdProyecto();

        long siguienteIdTarea();
    }

    /** Para las pruebas. Un mapa y dos contadores. */
    public static final class RepositorioEnMemoria implements Repositorio {
        private final Map<Long, Proyecto> proyectos = new HashMap<>();
        private long siguienteProyecto = 1;
        private long siguienteTarea = 1;

        @Override
        public Proyecto porId(long id) {
            return proyectos.get(id);
        }

        @Override
        public Proyecto guardar(Proyecto proyecto) {
            proyectos.put(proyecto.id, proyecto);
            return proyecto;
        }

        @Override
        public long siguienteIdProyecto() {
            return siguienteProyecto++;
        }

        @Override
        public long siguienteIdTarea() {
            return siguienteTarea++;
        }
    }
}
